platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/ConfigurationMigrationManager.java

```diff
@@ -60,130 +61,220 @@
-            LOGGER.info("Failure to export, internal error occurred", e);
-            throw new UnexpectedMigrationException("Export failed", e);
-        }
+    public Collection<MigrationWarning> doExport(String exportDirectory) {
+        Validate.notNull(exportDirectory, "invalid null export directory");
+        final MigrationReport report = doExport(Paths.get(exportDirectory));
 
-        return migrationWarnings;
+        report.verifyCompletion(); // will throw error if it failed
+        return report.warnings()
+                .collect(Collectors.toList());
     }
 
-    public Collection<MigrationWarning> export(@NotNull String exportDirectory)
-            throws MigrationException {
-        notNull(exportDirectory, "Export directory cannot be null");
-
-        return export(Paths.get(exportDirectory));
+    @Override
+    public MigrationReport doExport(Path exportDirectory) {
+        return doExport(exportDirectory, Optional.empty());
     }
 
     @Override
-    public Collection<Describable> getOptionalMigratableInfo() {
-        return ImmutableList.copyOf(dataMigratables);
+    public MigrationReport doExport(Path exportDirectory, Consumer<MigrationMessage> consumer) {
+        Validate.notNull(consumer, "invalid null consumer");
+        return doExport(exportDirectory, Optional.ofNullable(consumer));
     }
 
-    private Collection<MigrationWarning> exportMigratable(Migratable migratable,
-            Path exportDirectory) {
-        Stopwatch stopwatch = null;
+    @Override
+    public Collection<MigrationWarning> doImport(String exportDirectory) {
+        Validate.notNull(exportDirectory, "invalid null export directory");
+        final MigrationReport report = doImport(Paths.get(exportDirectory));
 
-        if (LOGGER.isDebugEnabled()) {
-            stopwatch = Stopwatch.createStarted();
-        }
+        report.verifyCompletion(); // will throw error if it failed
+        return report.warnings()
+                .collect(Collectors.toList());
+    }
 
-        MigrationMetadata migrationMetadata = migratable.export(exportDirectory);
+    @Override
+    public MigrationReport doImport(Path exportDirectory) {
+        return doImport(exportDirectory, Optional.empty());
+    }
 
-        if (LOGGER.isDebugEnabled() && stopwatch != null) {
-            LOGGER.debug("Export time: {}",
-                    stopwatch.stop()
-                            .toString());
-        }
+    @Override
+    public MigrationReport doImport(Path exportDirectory, Consumer<MigrationMessage> consumer) {
+        Validate.notNull(consumer, "invalid null consumer");
+        return doImport(exportDirectory, Optional.of(consumer));
+    }
 
-        return migrationMetadata.getMigrationWarnings();
+    @VisibleForTesting
+    ImportMigrationManagerImpl delegateToImportMigrationManager(MigrationReportImpl report,
+            Path exportFile) {
+        final ImportMigrationManagerImpl mgr = new ImportMigrationManagerImpl(report,
```

Comment:
❓ Don't you want to wrap this impl in a try-with-resources?

Reply:
too bad, you can see the reply on the first PR. We can talk about it when you are back. See previous reply on the subject about closing input streams.